dask_jobqueue/job.py

```diff
@@ -0,0 +1,394 @@
+                "{!r}".format(self.job_id_regexp)
+            )
+            raise ValueError(msg)
+
+        return job_id
+
+    async def close(self):
+        logger.debug("Stopping worker: %s job: %s", self.name, self.job_id)
+        self._close_job(self.job_id)
+
+    @classmethod
+    def _close_job(cls, job_id):
+        if job_id:
+            cls._call(shlex.split(cls.cancel_command) + [job_id])
+
+    @staticmethod
+    def _call(cmd, **kwargs):
+        """ Call a command using subprocess.Popen.
+
+        This centralizes calls out to the command line, providing consistent
+        outputs, logging, and an opportunity to go asynchronous in the future.
+
+        Parameters
+        ----------
+        cmd: List(str))
+            A command, each of which is a list of strings to hand to
+            subprocess.Popen
+
+        Examples
+        --------
+        >>> self._call(['ls', '/foo'])
+
+        Returns
+        -------
+        The stdout produced by the command, as string.
+
+        Raises
+        ------
+        RuntimeError if the command exits with a non-zero exit code
+        """
+        cmd_str = " ".join(cmd)
+        logger.debug(
+            "Executing the following command to command line\n{}".format(cmd_str)
+        )
+
+        proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs
+        )
+
+        out, err = proc.communicate()
+        if six.PY3:
+            out, err = out.decode(), err.decode()
+        if proc.returncode != 0:
+            raise RuntimeError(
+                "Command exited with non-zero exit code.\n"
+                "Exit code: {}\n"
+                "Command:\n{}\n"
+                "stdout:\n{}\n"
+                "stderr:\n{}\n".format(proc.returncode, cmd_str, out, err)
+            )
+        return out
+
+
+class JobQueueCluster(SpecCluster):
+
+    def __init__(
+        self,
+        n_workers=0,
+        Job: Job = None,
+        # Cluster keywords
+        loop=None,
+        security=None,
+        silence_logs=False,
+        name=None,
+        asynchronous=False,
+        # Scheduler keywords
+        interface=None,
+        protocol="tcp://",
+        dashboard_address=":8787",
+        # Job keywords
+        **kwargs
+    ):
+        if Job is None:
+            raise ValueError(
+                "You must provide a Job type like PBSJob, SLURMJob, "
+                "or SGEJob with the Job= argument."
+            )
+
+        scheduler = {
+            "cls": Scheduler,  # Use local scheduler for now
+            "options": {
+                "protocol": protocol,
+                "interface": interface,
+                "dashboard_address": dashboard_address,
+                "security": security,
+            },
+        }
+        kwargs["interface"] = interface
+        kwargs["protocol"] = protocol
+        kwargs["security"] = security
+        worker = {"cls": Job, "options": kwargs}
```

Comment:
That's probably here we should add other options needed by adaptive, e.g. cores and memory (see https://github.com/dask/distributed/pull/2974).